### ISSUE TITLE
Spot and fiducial diameters should be stored somewhere

### ISSUE THREAD

**Issue by LTLA** [drighelli/SpatialExperiment#25]
Probably would go nicely in the metadata of the `imgData` DF, provided this is well-documented in `?read10xVisium`.

**Comment by HelenaLC** [comment]
Sorry, but what is this? Are these specific to each image, or the whole sample? Is this explained somewhere on the 10x website?

**Comment by LTLA** [comment]
Floating around in `scalefactors_json.json`. Yes, probably somewhere in `int_metadata`.

**Comment by HelenaLC** [comment]
Thinking about this more... the `scalefactors_json.json` file is unique to each sample. So just thinking out loud...
- Currently, we store a scale factor for each sample-image, so this is redundant if there are multiple images of the same resolution.
- The spot/fiducial diameter is experiment-wide, but we cannot be sure that each SPE will contain only data from the same technology/generation etc. 
- So, I wonder if we should store this in a different way... we have a list of scale factors for each sample, but currently store a single scale factor for each image. 
Thoughts?

**Comment by LTLA** [comment]
Agree that redundancy is fine; there aren't going to be any changes to the scale factor in the course of the analysis, so we don't have to worry about keeping things in sync. And it also ensures that, e.g., if you replace a image with one of different resolution, you have the ability to change the scale factor for that image only, rather than having to worry about rematching.

You could also store the fiducials in some per-image metadata. Maybe even inside the `SpatialImage()` itself, if we make the base class extend from `Annotated`; then we get a `metadata()` field for free.

**Comment by drighelli** [comment]
I agree with keeping the whole scale factors associated with each image.

Since you are working on the SpatialImage, do you want to do this @HelenaLC ?

**Comment by HelenaLC** [comment]
Okay, maybe I wasn't all too clear with what I meant... 
1. Storing the scale factor inside the `SpatialImage` would be essentially the same as what we have now, where we store it "next" to the image.
2. This doesn't solve the "issue" that we have 4 scale factors per sample, but only 1 scale factor per image, and 2 scale factors that are sample-, not image-specific.

Proposal:
An intuitive way of handling this for me is to follow along with how the data comes from Cell Ranger... E.g. `imgData` could look like this, which is fairly compact IMO:

```
         images  scaleFactors
sample1  <list>  <list>
sample2  <list>  <list>
```
where each row is a sample, and `images` a list of `SpatialImages`, which could in turn store the scale factor specific to that image (to omit having to match this downstream), i.e. 1 of the entries in `scaleFactors`.

`height` & `width` are gone (we get this from `dim(SpatialImage)`), `sample_id`s become rownames, and `image_id`s could be simply the names of the `images` list.

**Comment by LTLA** [comment]
> 2\. This doesn't solve the "issue" that we have 4 scale factors per sample, but only 1 scale factor per image, and 2 scale factors that are sample-, not image-specific.

I don't understand the problem here. Why is the scaling factor sample-specific? Surely it's image-specific; the fact that you have multiple images of the same resolution for a given sample is irrelevant. I assume you're talking about the 1 high + 3 low-res images (or maybe the other way around, I don't remember) associated with each sample.

> An intuitive way of handling this for me is to follow along with how the data comes from Cell Ranger... E.g. `imgData` could look like this, which is fairly compact IMO:

This makes the `images` harder to search for and get to, because now you need to go inside each list to, e.g., find all tissue section images. You can also imagine future applications where extra image metadata fields (e.g., z-index, strain method) could be added to a DF easily, while relying on names would be a pain. And I also don't understand why `scaleFactors` is a separate list, and not stored with the `SpatialImage`.

**Comment by HelenaLC** [comment]
I meant that the scaling factor file (4 scaling factors) is sample-specific. Agree that each image has 1 scaling factor associated with it; this can go into the `SpatialImage` `metadata` if we inherit from `Annotated`. However, I'm still not sure where to put the spot/fiducial diameter - these are sample-specific, not image-specific. If it were always the case that all samples from from the sample tech generation, these could go into `(int_)metadata`. But to be general, each sample could come from a different slide and have different diameters. Where do you suppose these should go? Maybe we could have something general like...
```
sample_id image_id images metadata
```
where images: list of SpatialImages with one scaleFactor each, and metadata: a list of anything else we/the user wants to put in there?

**Comment by LTLA** [comment]
> However, I'm still not sure where to put the spot/fiducial diameter - these are sample-specific, not image-specific.

I wouldn't worry about it. As we discussed, redundancy is fine; these things won't change during the course of the analysis.

> Anyways, for now, I can put drop `scaleFactor` column from the `imgData` and store the scale factors inside the `SpatialImage`?

I think that's fine.

**Comment by lmweber** [comment]
Will close this issue for now. This issue may come up again in the future if we have different spot diameters in a new version of the Visium platform (e.g. Visium HD), but this is unlikely to be an issue within the next year or longer.

**Comment by HelenaLC** [comment]
@lmweber @drighelli trying to work off some issues... have we come to a conclusion regarding this? Here some options for discussion:
```r
sample_id image_id data metadata
# where 'data' has the 'SpatialImage' and
# 'metadata' everything that's in the .json
# OR store the scale factor in the SPI?

sample_id image_id data scaleFactor metadata
# same as above, but a separate slot for 'scaleFactor'
```

**Comment by HelenaLC** [comment]
Could you clarify, in this layout, what would go in `metadata`? I think those columns already cover the whole `.json`, no?

Regarding calculating diameters: I personally would prefer offering simple wrappers to do this (i.e., extract resolution and diameters and compute), which could be called by Visium-specific plotting functions. The reason being that these are highly tech-specific parameters that wouldn't exist for e.g. STEREO-seq or ISH-based platforms. But that's just my preference in order to not additionally clutter the `imgData` (as I think we'd want to keep both the original as well as the resolution-dependent values).

**Comment by LTLA** [comment]
Having considered the options, most sustainable way to go seems to be a `metadata` column containing a `List`. This can be treated as a giant dump of image-specific miscellany, which means that readers for other technologies can easily store whatever they want in there. So at the very least, we're not losing any technology-specific information.

The problem is when we want to provide guarantees to downstream methods about the existence of certain fields in this `metadata` column. We wouldn't be able to do that with a loose `metadata` column, but perhaps we can provide some recommendations, e.g., "`metadata`  should contain `fidicual_diameter` and `spot_diameter` for Visium data".

If we want to give hard guarantees, then we could create `VisiumExperiment` subclass that guarantees the existence of a `fiducial_diameter` column, etc. This seems like overkill and increases fragmentation of the class hierarchy.

**Comment by LTLA** [comment]
Nudging this. Putting these parameters somewhere would be helpful for downstream applications.

**Comment by lmweber** [comment]
Thanks for the reminder.

Yes, maybe putting it all in `metadata()` in this way, along with some recommended guidelines for specific platforms (especially Visium) is a good compromise between generality and platform-specificity.

**Comment by drighelli** [comment]
Yes, I also think that the best way is to store them inside the `metadata` by indexing per `sample_id`

**Comment by lmweber** [comment]
Ok thanks, I'll try this out with an example

**Comment by lmweber** [comment]
Rebasing `SpatialImage` classes on `Annotated` in PR #122 now enables storing this in `metadata`.

To do: update `read10xVisium()` and add examples as discussed in PR #122 .
